File: 1 Kyu/1 kyu - Tiny Three-Pass Compiler.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <stack>
//#include <regex>
#include <functional>
#include <chrono>

using namespace std;
using func = function<double(double,double)>;

map<string,int> order {{"+", 1},{"-",1},{"*",2},{"/",2}};
map<string, func> oper {{"+", plus<double>()},{"-", minus<double>()},
              {"*", multiplies<double>()}, {"/", divides<double>()}};

struct AST {
    string op;
    int n;
    AST *a, *b;

    AST (const string &label = "", const int &num = 0) : op (label), n (num), a (nullptr), b (nullptr) {}
    AST (const string &label, AST *arg1, AST *arg2) : op (label), a (arg1), b (arg2) {}
};
/////////////////////////////////Assert/////////////////////////////////////////
class Assert {
    public :
        template<class T> static void That (const T& a, const T& b) {
            if (a != b) {
                cout << "actual : " << a << " expected : " << b;
                cout << endl;
            }
        }
        template<class T> static void That (const vector<T> &a, const vector<T> &b) {
            if (a != b) {
                cout << "actual : ";
                for (auto &it : a) cout << it << " ";

                cout << " expected : ";
                for (auto &it : b) cout << it << " ";

                cout << endl;
            }
        }
};
template<class T> T Equals (const T& entry) { return entry;}
template<class T> T EqualsContainer (const T& entry) { return entry;}
void Test ();
////////////////////////////////////////////////////////////////////////////////
class Timer {
    //using time = chrono::steady_clock;
    private :
        chrono::steady_clock::time_point alpha, end;
        chrono::duration<double> elapsed;
        uint64_t index;

        void update () {
            end = chrono::steady_clock::now ();
            elapsed = end - alpha;
        }
    public :
        Timer () {
            alpha = chrono::steady_clock::now ();
            index = 0;
        }
        void start () { alpha = chrono::steady_clock::now ();}
        void stop () { update();}
        void get_duration () {

            cout << "\nDuration "
                      <<fixed<< elapsed.count()
                      << " ms" << endl;
        }
        bool running (double total) {
            update();
            index++;
            if (elapsed.count() < total) return true;
                cout << "index :: " << index << endl;
            return false;
        }
};
/////////////////////////////////tools//////////////////////////////////////////
#define Bin(op,a,b) (new AST (#op, (a), (b)))
#define Arg(op,n) (new AST (#op, (n)))

bool equals (const AST &lhs, const AST &rhs) {
  return lhs.op == rhs.op && lhs.n == rhs.n && (lhs.a && rhs.a ? equals (*lhs.a, *rhs.a) : true) && (lhs.b && rhs.b ? equals (*lhs.b, *rhs.b) : true);
}
bool equals (const AST *lhs, const AST *rhs) { return equals (*lhs, *rhs); }
int simulate (const vector<string> &assembly, const vector<int> &argv) {
    int r0 = 0, r1 = 0;
    stack<int> istack;

    for (auto &ins : assembly) {
        //cout << r0 << " ";
        string code = ins.substr (0, 2);
        //cout << ins << " ";
             if (code == "IM") { r0 = stoi (ins.substr (3)); }
        else if (code == "AR") { r0 = argv.at (stoi (ins.substr (3))); }
        else if (code == "SW") { swap (r0, r1); }
        else if (code == "PU") { istack.push (r0); }
        else if (code == "PO") { r0 = istack.top (); istack.pop (); }
        else if (code == "AD") { r0 += r1; }
        else if (code == "SU") { r0 -= r1; }
        else if (code == "MU") { r0 *= r1; }
        else if (code == "DI") { r0 /= r1; }
    }
    return r0;
}

class Display {
    public :
        template<class T> static void vect (const vector<T> &now) {
            for (auto &it : now) cout << it << " ";
            cout << endl;
        }
        static void node (const AST *now) {
            if (oper[now->op]) {
                cout << "Bin(" << now->op << ", ";
            } else {
                cout << "Arg(" << now->op << "," << now->n;
            }
        }
        static void tree (const AST *now) {
            if (now != nullptr) {
                node (now);
                tree (now->a);
                tree (now->b);
                cout << ")";
            }
        }
};

////////////////////////////////////////////////////////////////////////////////
class Compiler {
    private :
        map<string, int> args;

        bool isnum (const string &ex) {

            if (oper[ex]) return false;

            for (auto &it : ex) {
                if (isalpha (it)) return false;
                if (it == '(' || it == ')') return false;
            }
            return true;
        }
        AST *getstk (stack<AST *> &S) {   // get value from stack && delete stack
            AST *now = S.top();
            S.pop();
            return now;
        }
        stack<AST *> postorder (AST *root) {      // post order traversal

            AST *temp = root;
            stack<AST *> s1, s2;
            s1.push (temp);

            while (!s1.empty()) {
                temp = getstk (s1);
                s2.push (temp);

                if (temp->a) s1.push(temp->a);
                if (temp->b) s1.push(temp->b);
            }

            return s2;
        }
        AST *mktree (const vector<string> &expr) { // make AS tree
            if (!expr.size()) throw runtime_error ("Empty expression");

            bool running = true;
            vector<string>::const_iterator it = expr.begin();
            stack<AST *> tree, ops;

            while (running) {
                string tile = *it;

                if (tile == "(") {
                    it++;
                    auto nxt = find (it, expr.end(), ")");

                    AST *sub = mktree ({it, nxt});
                    tree.push (sub);

                    it = nxt;
                } else if (isnum (tile)) {
                    tree.push (new AST ("imm", stoi (tile)));
                } else if (args[tile]) {
                    tree.push ((new AST ("arg", args[tile] - 1)));
                } else if (oper[tile]) {
                    AST *temp = new AST (tile);

                    while (!ops.empty () && order[ops.top()->op] >= order[tile]) {
                        AST *op = getstk (ops);
                        op->b = getstk (tree), op->a = getstk (tree);
                        tree.push (op);
                    }

                    ops.push (temp);
                } else {

                    throw::logic_error ("invalid identifier");
                }

                it++;
                if (it - expr.begin() >= expr.size()) running = false;
            }

            while (!ops.empty ()) {
                AST *op = getstk (ops);
                op->b = getstk (tree), op->a = getstk (tree);
                tree.push (op);
            }

            return tree.size() ? tree.top() : nullptr;
          }

    public :
        vector<string> compile (string program) {
            return pass3 (pass2 (pass1 (program)));
        }
        vector<string> tokenize (string program) {
            // Turn a program string into a vector of tokens.  Each token
            // is either '[', ']', '(', ')', '+', '-', '*', '/', a variable
            // name or a number (as a string)
            //static regex re ("[-+*/()[\\]]|[A-Za-z]+|\\d+");
            //sregex_token_iterator it (program.begin (), program.end (), re);
            //return vector<string> (it, sregex_token_iterator ());
            string::iterator it = program.begin();
            vector<string> tok;

            while (*it) {
                if (*it != ' ') tok.push_back ({*it});
                it++;
            }
            return tok;
        }

        AST *pass1 (string program) {     // Returns an un-optimized AST
            auto tokens = tokenize (program);
            vector<string>::iterator it = tokens.begin();

            int i = 1;
            for (it++ ; *it != "]"; it++)
                  args[*it] = i++;

            AST *ASTree = mktree ({it + 1, tokens.end()});

            return ASTree;
        }
        AST *pass2 (AST *ast) {           // Returns an AST with constant expressions reduced

            if (ast->op == "imm" || ast->op == "arg") return ast;
            ast->a = pass2 (ast->a), ast->b = pass2 (ast->b);

            if (ast->a->op == "imm" && ast->b->op == "imm") {
                ast->n = oper[ast->op] (ast->a->n, ast->b->n);
                ast->op = "imm";
                ast->a = nullptr, ast->b = nullptr;
            }
            /*
            stack<AST *> s1 = postorder (ast);
            while (!s1.empty()) {
                reduce (getstk (s1));
            }
            */

            return ast;
        };
        vector<string> pass3 (AST *ast) { // Returns assembly instructions

            vector<string> Asm, pref {"PO", "SW", "PO"};
            stack<AST *> tree = postorder(ast);

            while (!tree.empty()) {
                AST *node = getstk (tree);

                if (node->op == "imm") { Asm.push_back ("IM " + to_string (node->n));
                } else if (node->op == "arg") { Asm.push_back ("AR " + to_string (node->n));
                } else {
                    Asm.insert (Asm.end(), pref.begin(), pref.end());

                    if (node->op == "+") { Asm.push_back("AD");
                    } else if (node->op == "-") { Asm.push_back("SU");
                    } else if (node->op == "*") { Asm.push_back("MU");
                    } else if (node->op == "/") { Asm.push_back("DI");
                    }
                }

                Asm.push_back ("PU");
            }

            return Asm ;
          }
};

int main () {
    Timer clock;

    Test ();

    clock.stop();
    clock.get_duration();
}

void Test () {

      { // AST_creation
        AST *ast = Bin(+,Arg(imm,1),Arg(imm,2));
        Assert::That (false, Equals (ast == 0));

        Compiler *compiler = new Compiler ();
        Assert::That (false, Equals (compiler == 0));
      }


     {
        string prog = "[ x y z ] ( 2*3*x + 5*y - 3*z ) / (1 + 3 + 2*2)";
        vector<string> tokens1 ({"[", "x", "y", "z", "]", "(", "2", "*", "3", "*", "x", "+", "5", "*", "y", "-", "3", "*", "z", ")", "/", "(", "1", "+", "3", "+", "2", "*", "2", ")"});

        Compiler compiler;
        auto tokens = compiler.tokenize (prog);
        Assert::That (tokens, EqualsContainer (tokens1));
    }

    {
        string prog = "[] 1";
        AST *ast1 = Arg(imm,1);

        Compiler compiler;
        AST *ast = compiler.pass1 (prog);
        Assert::That (true, Equals (equals (*ast, *ast1)));
    }

    {
        string prog = "[] 1 - 2";
        AST *ast1 = Bin(-,Arg(imm,1),Arg(imm,2));

        Compiler compiler;
        AST *ast = compiler.pass1 (prog);
        Assert::That (true, Equals (equals (*ast, *ast1)));
    }

    {
        string prog = "[x] x";
        AST *ast1 = Arg(arg,0);

        Compiler compiler;
        AST *ast = compiler.pass1 (prog);
        Assert::That (true, Equals (equals (*ast, *ast1)));
    }

    {
        string prog = "[x y z] z";
        AST *ast1 = Arg(arg,2);

        Compiler compiler;
        AST *ast = compiler.pass1 (prog);
        Assert::That (true, Equals (equals (*ast, *ast1)));
    }

    {
        string prog = "[x] x+2*5";
        AST *ast1 = Bin(+,Arg(arg,0),Bin(*,Arg(imm,2),Arg(imm,5)));

        Compiler compiler;
        AST *ast = compiler.pass1 (prog);
        Assert::That (true, Equals (equals (*ast, *ast1)));
    }

    {
        string prog = "[ x y z ] ( 2*3*x + 5*y - 3*z ) / (1 + 3 + 2*2)";
        AST *ast1 = Bin(/,Bin(-,Bin(+,Bin(*,Bin(*,Arg(imm,2),Arg(imm,3)),Arg(arg,0)),Bin(*,Arg(imm,5),Arg(arg,1))),Bin(*,Arg(imm,3),Arg(arg,2))),Bin(+,Bin(+,Arg(imm,1),Arg(imm,3)),Bin(*,Arg(imm,2),Arg(imm,2))));

        Compiler compiler;
        AST *ast = compiler.pass1 (prog);
        Assert::That (true, Equals (equals (*ast, *ast1)));
    }

    {
        string prog = "[x] x+2*5";
        AST *ast1 = Bin(+,Arg(arg,0),Arg(imm,10));

        Compiler compiler;
        AST *ast = compiler.pass2 (compiler.pass1 (prog));
        Assert::That (true, Equals (equals (*ast, *ast1)));
    }

    {
        Compiler compiler;
        AST *ast = compiler.pass2 (Bin(-,Arg(imm,10),Arg(imm,2)));
        Assert::That (true, Equals (equals (ast, Arg(imm,8))));
    }

    {
        Compiler compiler;
        AST *ast = compiler.pass2 (Bin(/,Arg(imm,10),Arg(imm,2)));
        Assert::That (true, Equals (equals (ast, Arg(imm,5))));
    }

    {
        string prog = "[ x y z ] ( 2*3*x + 5*y - 3*z ) / (1 + 3 + 2*2)";
        AST *ast1 = Bin(/,Bin(-,Bin(+,Bin(*,Arg(imm,6),Arg(arg,0)),Bin(*,Arg(imm,5),Arg(arg,1))),Bin(*,Arg(imm,3),Arg(arg,2))),Arg(imm,8));

        Compiler compiler;
        AST *ast = compiler.pass2 (compiler.pass1 (prog));
        Assert::That (true, Equals (equals (*ast, *ast1)));
    }

    {
        string prog = "[ x y z ] ( 2*3*x + 5*y - 3*z ) / (1 + 3 + 2*2)";

        Compiler compiler;
        auto c = compiler.compile (prog);
        Assert::That (simulate (c, {5, 4, 1}), Equals (5));
    }

    {
        string prog = "[ x y z ] ( 2*3*x + 5*y - 3*z ) / (1 + 3 + 2*2)";
        Compiler compiler;
        auto c = compiler.compile (prog);
        Assert::That (simulate (c, {4, 0, 0}), Equals (3));
        Assert::That (simulate (c, {4, 8, 0}), Equals (8));
        Assert::That (simulate (c, {4, 8, 16}), Equals (2));
    }

    {
        string prog = "[ x y z ] x - y - z + 10 / 5 / 2 - 7 / 1 / 7";
        Compiler compiler;
        auto c = compiler.compile (prog);
        Assert::That (simulate (c, {5, 4, 1}), Equals (0));
    }
    /*
    */
};
//////////////////////////////Arkive//////////////////////////////////////////
/*
void reduce (AST *curr) {         // reduce constant expression if possible

    if (oper[curr->op]) {

        AST *a = curr->a, *b = curr->b;

        if (a->op == "imm" && b->op == "imm") {
            curr->n = oper[curr->op] (a->n, b->n);
            curr->op = "imm";
            curr->a = nullptr, curr->b = nullptr;
            free (a), free (b);
        }
    }

}

int probe (AST* node) {           // search for max level
    if (node == nullptr) return -1;
    return max (probe (node->a), probe (node->b)) + 1;
}
stack<AST *> postorder (AST *root) {      // post order reduction

    stack<AST *> s1,s2;
    AST *temp = root;
    s1.push (temp);

    while (!s1.empty()) {
        temp = s1.top();
        s1.pop();

        s2.push (temp);

        if (temp->b) s1.push(temp->b); // Push the right child of the top element
        if (temp->a) s1.push(temp->a); // Push the left child of the top element
    }

    return s2;
}
void reach (AST *node, int depth, int dest) { //  recursion reduction

    if (node != nullptr) {
        if (depth == dest) reduce (node);

        reach (node->a, depth + 1, dest);
        reach (node->b, depth + 1, dest);
    }
}
*/

File: 1 Kyu/1 kyu - Tiny Three-Pass Compiler.rs
extern crate regex;
use regex::Regex;

#[derive(Clone,Debug)]
enum Ast {
    UnOp (String, i32),
    BinOp (String,Box<Ast>,Box<Ast>),
}

struct Compiler {
    args: Vec<String>,
}

impl Compiler {
    fn new() -> Compiler {
        Compiler {
            args: Vec::new(),
        }
    }

    fn getid (&self, ast : &Ast) -> String {

        if let Ast::BinOp(id, _a, _b) = ast {
            return id.to_string();
        } else if let Ast::UnOp(id, _val) = ast {
            return id.to_string();
        } else {
            return "null".to_string()
        }
    }
    fn getval (&self, ast : &Ast) -> i32 {

        if let Ast::UnOp(_id, val) = ast {
            return *val;
        }
        0
    }

    fn order (&self, tok: &str) -> usize {
        if tok == "+" || tok == "-" { return 1 }
        if tok == "*" || tok == "-" { return 2 }
        0
    }
    fn postorder (&self, root: &Ast) -> Vec<Ast> {

        let mut s1:Vec<Ast> = vec![root.clone()];
        let mut s2:Vec<Ast> = vec![];

        while let Some (node) = s1.pop() {

            if let Ast::BinOp(_id, a, b) = &node {
                s1.push(*a.clone());
                s1.push(*b.clone());
            }
            s2.push(node);
        }

        s2
    }
    fn mktree (&self, code: &[String]) -> Ast {

        let size = code.len();
        let number = Regex::new("^-?[0-9]+$").unwrap();
        let mut it = 0;
        let mut tree:Vec<Ast> = Vec::new();
        let mut oper:Vec<String> = Vec::new();

        while it != size {
            let tile = &code[it];

            if let Some (index) = self.args.iter().position(|x| x == tile) {
                tree.push (Ast::UnOp("arg".to_string(), index as i32));
            } else if tile == "(" {
                let fst = it + 1;

                while code[it] != ")" {
                    it += 1;
                }
                tree.push(self.mktree(&code[fst..it]))
            } else if number.is_match(tile) { // tile.chars().all(char::is_numeric) == true
                let value = tile.parse::<i32>().unwrap();
                tree.push (Ast::UnOp("imm".to_string(), value));
            } else {

                while let Some(id) = oper.last() {
                    if self.order (id) >= self.order (tile) {
                        let b = tree.pop().unwrap();
                        let a = tree.pop().unwrap();
                        tree.push (Ast::BinOp(id.to_string(), Box::new(a), Box::new(b)));
                        oper.pop();
                    } else {
                        break
                    }
                }

                oper.push (tile.to_string());
            }

            it += 1;
        }

        while let Some(id) = oper.pop() {

            let b = tree.pop().unwrap();
            let a = tree.pop().unwrap();
            tree.push (Ast::BinOp(id, Box::new(a), Box::new(b)));
        }

        if let Some(node) = tree.pop() { return node } else { Ast::UnOp ("null".to_string(),-1) }
    }

    fn tokenize<'a>(&self, program : &'a str) -> Vec<String> {
        let token = Regex::new("[-+*/()\\[\\]]|[A-Za-z]+|\\d+").unwrap();
        token.captures_iter(program).map(|x| x[0].to_string()).collect::<Vec<_>>()
    }
    fn compile (&mut self, program : &str) -> Vec<String> {

        let ast = self.pass1(program);
        let ast = self.pass2(&ast);
        self.pass3(&ast)
    }
    fn pass1 (&mut self, program : &str) -> Ast {

        let code = self.tokenize(program);
        let mut it = 1;

        while code[it] != "]" {
            self.args.push (code[it].to_string());
            it += 1;
        }

        self.mktree (&code[it+1..code.len()])
    }
    fn pass2 (&mut self, ast : &Ast) -> Ast {

        if let Ast::BinOp(id, a, b) = ast {
            let a:Ast = self.pass2(a);
            let b:Ast = self.pass2(b);

            if self.getid(&a) == "imm" && self.getid(&b) == "imm" {
                let value:i32;

                match id as &str {
                    "+" => value = self.getval(&a) + self.getval(&b),
                    "-" => value = self.getval(&a) - self.getval(&b),
                    "*" => value = self.getval(&a) * self.getval(&b),
                    "/" => value = self.getval(&a) / self.getval(&b),
                     _  => panic!("unknown operator"),
                }

                return Ast::UnOp("imm".to_string(), value);
            } else {
                return Ast::BinOp(id.clone(), Box::new(a), Box::new(b));
            }
        }

        return ast.clone();
    }
    fn pass3 (&mut self, ast : &Ast) -> Vec<String> {

        let mut asm = Vec::new();
        let mut tree:Vec<Ast> = self.postorder(ast);

        while let Some (node) = tree.pop() {

            if let Ast::BinOp(id, _a, _b) = &node {

                asm.push("PO".to_string());
                asm.push("SW".to_string());
                asm.push("PO".to_string());

                match id as &str {
                    "+" => asm.push("AD".to_string()),
                    "-" => asm.push("SU".to_string()),
                    "*" => asm.push("MU".to_string()),
                    "/" => asm.push("DI".to_string()),
                     _  => panic!("Invalid instruction"),
                }
            } else if let Ast::UnOp(id, val) = &node {

                match id as &str {
                    "imm" => asm.push("IM ".to_string() + &val.to_string()),
                    "arg" => asm.push("AR ".to_string() + &val.to_string()),
                      _   => panic!("Invalid instruction"),
                }
            }

            asm.push ("PU".to_string());
        }

        asm
    }
}
///////////////////////////////////////////////////////////////////////////////////////////////////
fn shownode (now: &Ast) {

    if let Ast::BinOp(id, _a, _b) = now {
        print!("Bin({}", id);
    } else if let Ast::UnOp(id, val) = now {
        print!("Arg({} {}", id, val);
    } else {
        print!("(null");
    }

}
fn showtree (now: &Ast) {

    shownode (now);
    if let Ast::BinOp(_id, a, b) = now {
        showtree (a);
        showtree (b);
    }
    print!(")");
}

fn simulator() {
    assert_eq!(simulate(vec!["IM 7".to_string()], vec![3]), 7);
    assert_eq!(simulate(vec!["AR 1".to_string()], vec![1,2,3]), 2);
}
fn simulate (assembly : Vec<String>, argv : Vec<i32>) -> i32 {
    let mut r = (0, 0);
    let mut stack : Vec<i32> = vec![];

    for ins in assembly {
        let mut ws = ins.split_whitespace();
        match ws.next() {
            Some("IM") => r.0 = i32::from_str_radix(ws.next().unwrap(), 10).unwrap(),
            Some("AR") => r.0 = argv[i32::from_str_radix(ws.next().unwrap(), 10).unwrap() as usize],
            Some("SW") => r = (r.1,r.0),
            Some("PU") => stack.push(r.0),
            Some("PO") => r.0 = stack.pop().unwrap(),
            Some("AD") => r.0 += r.1,
            Some("SU") => r.0 -= r.1,
            Some("MU") => r.0 *= r.1,
            Some("DI") => r.0 /= r.1,
            _ => panic!("Invalid instruction encountered"),
        }
    }
    r.0
}
///////////////////////////////////////////////////////////////////////////////////////////////////

fn main() {

    let program = "[ x y z ] ( 2*3*x + 5*y - 3*z ) / (1 + 3 + 2*2)";

    let mut asm = Compiler::new();
    let tree = asm.pass1 (program);
    let tree = asm.pass2 (&tree);
    let code = asm.pass3 (&tree);

    let res = simulate (code, vec![5, 4, 1]);

    print! ("{} => {}",program, res);

    /*
    println!("AST1: {:?}", tree);
    //showtree(&tree);
    */

    print!("\n");
}
